workflow-tracing/src/main/java/com/squareup/workflow1/diagnostic/tracing/TimelineWorkflowInterceptor.kt

```diff
@@ -0,0 +1,308 @@
+package com.squareup.workflow1.diagnostic.tracing
+
+import com.squareup.workflow1.BaseRenderContext
+import com.squareup.workflow1.Snapshot
+import com.squareup.workflow1.Workflow
+import com.squareup.workflow1.WorkflowAction
+import com.squareup.workflow1.WorkflowIdentifier
+import com.squareup.workflow1.WorkflowIdentifierType.Snapshottable
+import com.squareup.workflow1.WorkflowIdentifierType.Unsnapshottable
+import com.squareup.workflow1.WorkflowInterceptor
+import com.squareup.workflow1.WorkflowInterceptor.RenderContextInterceptor
+import com.squareup.workflow1.WorkflowInterceptor.WorkflowSession
+import com.squareup.workflow1.applyTo
+import com.squareup.workflow1.diagnostic.tracing.TimelineWorkflowInterceptor.RenderingState.None
+import com.squareup.workflow1.diagnostic.tracing.TimelineWorkflowInterceptor.RenderingState.Rendering
+import com.squareup.workflow1.diagnostic.tracing.TimelineWorkflowInterceptor.RenderingState.Snapshotting
+import com.squareup.workflow1.identifier
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlin.reflect.KClass
+import kotlin.reflect.KType
+
+/**
+ * A [WorkflowInterceptor] that just prints all method calls using [log].
+ */
+public open class TimelineWorkflowInterceptor : WorkflowInterceptor {
+
+  private data class Edge(
+    val from: String,
+    val to: String,
+    val label: String?
+  )
+
+  private sealed interface RenderingState {
+    object None : RenderingState
+    data class Rendering(val root: WorkflowSession): RenderingState
+    data class Snapshotting(val root: WorkflowSession): RenderingState
+  }
+
+  private val stack = ArrayDeque<String>()
+  private val causalChain = mutableListOf<Edge>()
+  private var nextRenderingCause: String? = null
+  private var renderingState: RenderingState = None
+  private var enqueuedActions = 0
```

Comment:
When you get a chance can you document what each of these are keeping track of?

Reply:
if I don't throw the whole thing away, yes :)